Add a new mode to use most of the GPU support without GPUs

e-kayrakli opened the pull request:
This PR makes significant changes to GPU-related Chapel environment and adds a new mode to use the GPU locale model without GPUs.

Primarily resolves https://github.com/Cray/chapel-private/issues/4621
While there resolves https://github.com/chapel-lang/chapel/issues/21978
While there resolves https://github.com/Cray/chapel-private/issues/4480

Note: The PR message is evolving and it is not in its final state.

### Prelude: Environment/Script Changes
While some was necessary for this to work, some of the following are more for general improvements:
- `CHPL_GPU_CODEGEN` and `CHPL_GPU_RUNTIME` are now ignored with a warning if they're set
- `CHPL_GPU` can now be used to choose between `nvidia`, `amd` and `cpu`. `cpu` is the new mode added in this PR.
- `CHPL_GPU` is only a RUNTIME environment variable. i.e., `nvidia` and `amd` targets don't need separate compiler builds. Resolves https://github.com/chapel-lang/chapel/issues/21978.
- It is also not NOPATH, leading to separate directories for runtime builds for different vendors. Resolves https://github.com/chapel-lang/chapel/issues/21978.
- `CHPL_GPU_ARCH`, `CHPL_CUDA_LIBDEVICE_PATH`, `CHPL_CUDA_PATH` and `CHPL_ROCM_PATH` are now INTERNAL environment variable. i.e., they don't show up unless `--internal` was used. Resolves https://github.com/Cray/chapel-private/issues/4480.
- "tidy" filter in `printchplenv` now suppresses CUDA variables with AMD and vice versa. Resolves https://github.com/Cray/chapel-private/issues/4480
- `util/chplenv/printchplbuilds.py` now prints out builds differ based on `CHPL_GPU` and `CHPL_GPU_MEM_STRATEGY`
- `CHPL_GPU_MEM_STRATEGY` has its proper shortname -- leading to proper directory in the build path for the runtime.
- `GPU_TYPES` in `chpl_gpu` now is a dict that stores `gpu_type` objects instead of tuples
- move `llvm` check from `chpl_compiler.py` to `chpl_gpu.py` to avoid a circular dependence in scripts

The rest of the message is about the new `CHPL_GPU=cpu` mode.

### `CHPL_GPU=cpu`: Compiler Changes/Behavior
- Kernels are generated normally
- compile time assertOnGpu still works
- Only significant change is with gpu and non-gpu paths. With this new mode we generate:

```chapel
if runningOnGpuLocale() {
  launch_kernel(....);
}
{  // not an else block
  foreach ...
}
```

IOW, we don't put the original loop in an else block but in an anonymous one. We could flatten that block, but I wanted to keep things as close to real GPU codegen as possible. If we generate something in that block down the road, this'll help.

### `CHPL_GPU=cpu`: Runtime Changes/Behavior
- To adapt to new environment settings we now have `nvidia` and `amd` in names of dirs and files instead of `cuda` and `rocm`. I liked the previous ones a little bit better, but keeping them seemed like overkill. We'll adapt.
- Adds a new `cpu` directory as a sibling to `nvidia` and `amd` as needed
- `gpu-cpu.c` implements all runtime functions as appropriate while leaving some empty
- `assertOnGpu` in this mode generates a warning that can be suppressed by `--gpuNoCpuModeWarning` or `CHPL_GPU_NO_CPU_MODE_WARNING`
- `gpuClock` still calls `clock` but also generates a suppressible warning
- removes `chpl_gpu_get_device_count` from the runtime interface
- instead, number of devices is stored in a global
- `chpl-gpu.c` sets this based on what the implementation layer says for the number of devices available, and the new `CHPL_RT_NUM_GPU_PER_LOCALE`. This new environment variable can be used to limit number of devices used with "real" GPU configs. With `CHPL_GPU=cpu`, it can be set to any positive integer.

### `CHPL_GPU=cpu`: Module Changes/Behavior
- `syncThreads` is no-op in the new mode
- `createSharedArray` "works". In all likelihood, code that uses shared arrays with this mode will not run correctly. There's no easy way of implementing that with this mode.

### Other changes
- document the new mode, update technote w.r.t. other changes
- remove `studies.skipif`: was added in the early days of AMD support, now is not necessary.


Test:
- [x] gpu/native on `CHPL_GPU=nvidia`
- [x] gpu/native on `CHPL_GPU=amd`
- [x] gpu/native/noGpu on `CHPL_GPU=cpu`
- [x] `make check` on `CHPL_LOCALE_MODEL=flat`

## Open Questions/TODO:
- [x] Rename `CHPL_RT_NUM_DEVICES_PER_LOCALE` to `CHPL_RT_NUM_GPUS_PER_LOCALE` as suggested by @bradcray 
- [x] Is `none` the right name for this mode?
  - If not, coordinate with @stonea on the error message(s) in https://github.com/chapel-lang/chapel/pull/22170
- ~[ ] If we are using `none` for this mode, `printchplenv` should probably say `unset` for `CHPL_GPU_CODEGEN` and `CHPL_GPU_RUNTIME` by default without GPUs.~
- [x] Test if `CHPL_RT_NUM_DEVICES_PER_LOCALE` works with "real" GPU configs to limit number of GPU sublocales
- [x] Should we have nightly testing for this mode? (I think the answer is yes, but as a separate PR)
- [x] Cleanup commented code
- [x] Rebase and fix conflict

e-kayrakli commented:
@bradcray and @stonea -- pinging you as you both expressed interest in this mode before. It'd be great to get some input from you on Open Questions in the PR message.

stonea commented:
My main concern with the name `none` is that users might expect that to be a mode where no GPUs appear visible despite the fact that their system has GPUs. IOW a mode where we fix `here.gpus.size == 0`.

Having a mode like this might be important because:
- We may be able to generate more efficient code if we can assume that GPUs will never be used.
- User's might want to test that their code is robust enough to work on systems without GPUs without having to explicitly test on such a system.

For now the way to do this is just to set `CHPL_LOCALE_MODEL=flat` but at some point I would hope that the gpu locale model would become the default locale model. At that point I would expect user's would be more likely to consider `CHPL_GPU_CODEGEN` as the relevant variable for disabling GPU support. It seems "obvious" that `CHPL_GPU_CODEGEN=none` would mean "don't do GPU code generation".

A couple of alternatives I could imagine:

- Rename this layer to something else; like "sim".
- Keep the name 'none' but rename `CHPL_RT_NUM_GPUS_PER_LOCALE` to something like `CHPL_RT_NUM_SIMULATED_GPUS_PER_LOCALE` and have the default value for this be 0. The problem with this is we still might want to know that the user isn't interested in GPUs at compile-time.

e-kayrakli commented:
Good points. I like the name "none" better than "sim". It sounds more like a simulator than what this layer is about.

> Keep the name 'none' but rename CHPL_RT_NUM_GPUS_PER_LOCALE to something like CHPL_RT_NUM_SIMULATED_GPUS_PER_LOCALE and have the default value for this be 0. The problem with this is we still might want to know that the user isn't interested in GPUs at compile-time.

How about a variation of this:

- *Same as what we have here*: We keep the name as `none`. (+ `printchplenv` reports `unset` instead of `none` that it prints today)
- **New**: We also add a `--no-gpu` compiler flag. We wire it such that nothing that we do today at compile time happens at compilation time, including GPU kernel generation. The runtime will always report 0 devices, if the application is compiled this way. Just to draw a parallel with what we already have today; you can use `--local` in a `CHPL_COMM=gasnet`.
- *Same as what we have here*: `CHPL_RT_NUM_GPUS_PER_LOCALE=0` can be used at application launch time to result in similar execution behavior as `--no-gpu`, though by virtue of failing runtime checks and zeroing number of available devices. I guess currently this wouldn't work this way, ie, `assertOnGpu` would fail. But we can consider making that happen. We can also complement this with `--numgpus`/`-ng` default executable argument. 

Another high level comment in response to yours is that the mode that you are suggesting above feel like something we can achieve without rebuilding Chapel. Say, if we add `--no-gpu`, it would follow that we also add `CHPL_NO_GPU` environment variable that corresponds to that flag if you don't want to pass a flag each time you compile. That's what we do for all (most?) compiler flags today.

What do you think?

stonea commented:
I like the idea of having a `--no-gpu` compiler flag.  

The naming `CHPL_GPU_CODEGEN=none` still seems like it could lead to confusion and it doesn't seem self-evident what the difference between `CHPL_GPU_CODEGEN=unset` and `CHPL_GPU_CODEGEN=none` would be.

Imagine a world where the GPU locale model is the default.  It just seems weird to me that someone could write code like this:

```chpl
if here.gpus.size > 0 then writeln("I have GPUs on my system!");
```

And then set `CHPL_GPU_CODEGEN=none` and still have that message print. I'd want to say "hey I told you not to generate GPU code, why are you telling me I have GPUs?".

I think this mode is useful but I also want to make sure this isn't something users accidentally wander into.  The fact that `assertOnGpu` is still a warning alleviates my concern somewhat but before giving up on it I'd still like to see if we could come up with more self-evident naming.  How about one of these options:

* `CHPL_GPU_VENDOR=none`
* `CHPL_GPU_KIND=cpu`
* `CHPL_GPU_TARGET=cpu`

Again, if we really want to stick with `CHPL_GPU_CODEGEN=none` I'd be ok with that so long as users had to take some other action to "opt into" having these these "CPU bound GPU locales". Making `CHPL_RT_NUM_GPUS_PER_LOCALE` default to 0 seems like a step in the right direction; though again it kind of bothers me that I could set `CHPL_RT_NUM_GPUS_PER_LOCALE` to some non-zero number I could end up in a situation where I'm not using actual real HW GPUs. It's a mouthful (and maybe we can come up with something better) but how about something like `CHPL_RT_NUM_CPU_BOUND_GPUS_PER_LOCALE`?

This setting is only relevant when using the 'none' mode right?  What would the behavior of `--numgpus` be if run with 'CHPL_GPU_CODEGEN=cuda'?

bradcray commented:
> (+ printchplenv reports unset instead of none that it prints today)

I'm not really understanding the motivation for having `printchplenv` print something other than what the variable is set/inferred to (it seems like Andy has some similar hesitation here).

Andy's `sim` concept held some weight for me, though Engin's hesitations about potentially running on a GPU simulator also made sense.  Thinking about what it's really trying to say, I come up with things like `simulate-on-cpu` or `pretend-cpus-are-gpus` or `run-gpu-code-on-cpu`, all of which are too long but capture the execution mode well, I think?  Which makes me think of potentially just `cpu` (which seems similar to some of Andy's suggestions in his most recent comment).

bradcray commented:
Oh, and I agree that nightly testing for this mode would be nice.

e-kayrakli commented:
I may be missing something:

re:
> Again, if we really want to stick with CHPL_GPU_CODEGEN=none I'd be ok with that so long as users had to take some other action to "opt into" having these these "CPU bound GPU locales". 

and 

> I'm not really understanding the motivation for having printchplenv print something other than what the variable is set/inferred to (it seems like Andy has some similar hesitation here).

With this PR as-is, on a GPU-less system the default will be `unset`. You'll have to specifically set it to `none`. It'll be hard to wander into this mode.

However, I don't want to belabor this point as I see value in other directions that came up above.

I am most inclined towards dropping `CHPL_GPU_CODEGEN` and `CHPL_GPU_RUNTIME` and have something like `CHPL_GPU_TARGET=[nvidia|amd|cpu]`. However, (1) we probably would want `CHPL_TARGET_GPU` to match `CHPL_TARGET_CPU`. But even with that, Brad had some valid arguments against having `TARGET` in the name (we don't have `CHPL_HOST_GPU` after all, nor we have `CHPL_TARGET_TASKS` or `CHPL_TARGET_COMM`) when I proposed `CHPL_TARGET_GPU` about 1.5 years ago ([internal slack link](https://cray.slack.com/archives/C024Q0N3M5Y/p1638230659078800)). I don't think anything changed enough to argue against his arguments there. We might be looking towards `CHPL_GPU=[nvidia|amd|cpu]`. If `CHPL_GPU=cpu` reads weird, `CHPL_GPU_KIND=cpu` might be little less weird (to me they aren't not that different) and would be my second choice. FWIW, I see Brad's arguments against `TARGET` and wouldn't argue against them, though it wouldn't bother me too much if we use `CHPL_TARGET_GPU` here.

e-kayrakli commented:
Offline, I confirmed that Andy supports:

- Removing `CHPL_GPU_CODEGEN` and `CHPL_GPU_RUNTIME`
- Adding a `CHPL_GPU=[nvidia|amd|cpu]`

@bradcray -- I take your thumbs-up as supporting that direction, as well?

bradcray commented:
Yep, I like the simplicity of the approach compared to what we've had a lot, thanks!

stonea commented on runtime/src/chpl-gpu.c:
```diff
@@ -253,6 +286,12 @@ void* chpl_gpu_mem_realloc(void* memAlloc, size_t size,
+
+  // this is not really cool. Should we have a realloc in the impl layer?
+  if (cur_size == -1) {
+    return chpl_mem_realloc(memAlloc, size, description, lineno, filename);
```
Why did you add this?

e-kayrakli replied:
So, this is one of the earliest changes, and we probably have a better way of doing this right now.

The story goes: there is no reallocation in the CUDA API (probably neither in HIP). Here, this function does it in a home-grown way. It involves creating a new allocation, copying the old allocation into it and freeing the old one. But copying the old allocation requires a priori knowledge of the size of the old allocation. What CUDA API has instead is the ability to query that for a device pointer.

However, we don't have that ability available to us in this mode, so I made `chpl_gpu_get_alloc_size` to return -1 as a sentinel instead. Upon receiving that sentinel, we fall back to the normal memory layer. But maybe, we can check `GPU_RUNTIME_CPU` instead, now that I am thinking about this.

stonea commented on runtime/src/gpu/amd/gpu-amd.c:
```diff
@@ -128,29 +128,17 @@ static void chpl_gpu_impl_set_globals(hipModule_t module) {
-  #endif
-
-  // ROCM initialization
+void chpl_gpu_impl_init(int* num_devices) {
```
Thanks for cleaning on the AMD runtime and getting the `CHPL_GPU_DEBUG` lines out (since this is all handled by the vendor-neutral layer now right?).

e-kayrakli replied:
Right.

stonea commented on runtime/src/gpu/cpu/Makefile:
Funny git thinks this is a rename and modification to the file rather than a new file:

`runtime/src/gpu/rocm/Makefile → runtime/src/gpu/cpu/Makefile`

I'm guessing this is just git hiccuping; it looks like you have `amd/Makefile`, `cpu/Makefile` and `nvidia/Makefile` so I think we're fine.

e-kayrakli replied:
Yep.

stonea commented on runtime/src/gpu/cpu/gpu-cpu.c:
```diff
@@ -0,0 +1,158 @@
+
+void chpl_gpu_impl_init(int* num_devices) {
+  CHPL_GPU_DEBUG("Initializing none GPU layer.\n");
+  *num_devices = 1;
```
If you don't already please document in the tech note that the `CHPL_CPU` defaults to having one "virtual GPU" (or whatever terminology you want to call it).

stonea commented on runtime/src/gpu/cpu/gpu-cpu.c:
```diff
@@ -0,0 +1,158 @@
+static bool chpl_gpu_device_alloc = false;
+
+void chpl_gpu_impl_on_std_modules_finished_initializing(void) {
+  // The standard module has some memory that we allocate when we  are "on" a
```
This is called from a non-impl `chpl_gpu_on_std_modules_finished_initializing` function right?

I think I would just move the comment there are delete it in all of our _impl functions so we don't have this comment duplicated multiple times (too much risk it'll be modified in one copy and not another).

stonea commented on runtime/src/gpu/cpu/gpu-cpu.c:
```diff
@@ -0,0 +1,158 @@
+                                        int blk_dim_x,
+                                        int blk_dim_y,
+                                        int blk_dim_z,
+                                        int nargs, va_list args) {
```
This should never be called right? Do we want to assert?

e-kayrakli replied:
I think this can be called (although not tested) if you were to use the kernel launch primitive instead of going through user-facing means.

stonea commented on runtime/src/gpu/cpu/gpu-cpu.c:
```diff
@@ -0,0 +1,158 @@
+}
+
+void* chpl_gpu_impl_comm_async(void *dst, void *src, size_t n) {
+  chpl_memcpy(dst, src, n);
```
How liberal do we want to be with our warnings? We could have it warn here that with `CHPL_GPU=cpu` asynchronous communication will be made synchronous.

If nothing else document this in the tech note.

e-kayrakli replied:
I chose to document in the technote as a warning should be used in cases where the application behavior can change in a way that can generate incorrect results etc. Turning async memcpy into a blocking one is more innocent than that to me.

stonea commented on test/gpu/native/noGpu/SKIPIF:
```diff
@@ -0,0 +1 @@
+CHPL_GPU != cpu
```
Do we want to do nightly testing for "no gpu"? Probably? Anyway not something for this PR but create a GH issue unless you're convinced that we shouldn't do nightly testing in this mode.

e-kayrakli replied:
https://github.com/Cray/chapel-private/issues/4815

stonea commented on util/chplenv/chpl_compiler.py:
```diff
@@ -23,12 +23,6 @@ def validate_compiler(compiler_val, flag):
         if not os.path.isfile(comp_makefile):
             warning('Unknown compiler: "{0}"'.format(compiler_val))
 
-    if chpl_locale_model.get() == 'gpu' and flag == 'target':
```
Is it safe to remove this? Do we still error out if we're in this invalid state:

`CHPL_GPU_LOCALE_MODEL=gpu`, `CHPL_GPU!=cpu`, and `CHPL_TARGET_COMPILER!=llvm`?

e-kayrakli replied:
Yes, because this is now checked in `chpl_gpu.validate` instead. https://github.com/chapel-lang/chapel/pull/22172/files#diff-4c6fe7e6f01dadbf3eba9ac00e66c1c6546abaed9ed14bf5da46f278c7892bb4R283

stonea commented on util/cron/test-gpu-rocm.bash:
```diff
@@ -7,7 +7,6 @@ source $CWD/common-native-gpu.bash
 export CHPL_COMM=none
-export CHPL_GPU_CODEGEN=rocm
 export CHPL_LAUNCHER_PARTITION=amdMI60
 module load rocm
```
Maybe I'm going to far but I almost want to have this script fail if:

`printchplenv --all --simple | grep -o "CHPL_GPU=[^ ]*" | grep -o "[^=]*$"` != "amd"

I won't push on it, but update the cuda script if you do decide to add this check here.

e-kayrakli replied:
Is your fear based on the possibility of this falling back to `cpu` or by mistake run on a system with an NVIDIA gpu? I can put the equivalent (`export CHPL_GPU=amd`) both here and in the CUDA script? It is probably a better practice to be explicit in these scripts anyway.

stonea replied:
Yeah my fear is we're relying on `CHPL_GPU` to be correctly inferred to `amd` and in theory (though maybe somewhat unlikely) this could regress without us noticing. My knee-jerk defensive programming instinct says "add an assertion".

You're right that adding `export CHPL_GPU=amd` would assuage that fear.

Though I suppose it's also nice to rely/demonstrate the implicit behavior of `chplenv` setting the sensible default and then adding in this kind of "assertion" to verify it. We don't have any other real test for this feature so it'd be kind of nice to have this as one (even if it's kind of a sledgehammery way of doing it).

Anyway, if you don't buy my reasoning I'd be fine with just explicitly setting `CHPL_GPU=amd` there and `CHPL_GPU=cuda` in the cuda script and moving on.

e-kayrakli replied:
I don't disagree. But I don't find that assertion necessary. I'll go with explicit `CHPL_GPU` setting.

We set `CHPL_LOCALE_MODEL=gpu` explicitly. After that, by default `CHPL_GPU` could go `nvidia`/`amd` or `none`. I am confident that we wouldn't choose the wrong vendor. If we did, explosion would be seen from space. We could fall back to `none` if we can't find nvcc or hipcc, in which case build will fail.

You have to set `CHPL_GPU` explicitly to `cpu` to enter the new mode, and I find it very unlikely to happen. It might be set in a script that you include, I guess. In that case, setting `CHPL_GPU` explicitly should be enough, I suppose.

stonea commented on util/chplenv/chpl_gpu.py:
```diff
@@ -13,23 +14,50 @@ def _validate_cuda_version():
-#   depth of program within SDK folder
-#   Default GPU architecure for the vendor
-#   LLVM target
+class gpu_type:
```
Thank you for taking the initiative to refactor this into a class. Having everything packed as a tuple was starting to get unwieldy.

stonea commented on util/chplenv/chpl_gpu.py:
```diff
@@ -13,23 +14,50 @@ def _validate_cuda_version():
     if suggestNone:
         msg += " To avoid this issue, you can have GPU code run on the CPU "
-        msg += "by setting 'CHPL_GPU_CODEGEN=none'."
+        msg += "by setting 'CHPL_GPU=cpu'."
```
And thank you for remember to update this error message

stonea commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -898,29 +931,44 @@ static void generateGpuAndNonGpuPaths(const GpuizableLoop &gpuLoop,
+  // then relocate the loop
+  cpuBlock->insertAtHead(gpuLoop.loop()->remove());
+
+  // if not doing codegen, just add cpuBlock after the conditional
```
You need `if not doing GPU codegen` in: `// if not doing codegen, just add cpuBlock after the conditional`

e-kayrakli commented on compiler/llvm/clangUtil.cpp:
```diff
@@ -2365,6 +2365,8 @@ static std::string generateClangGpuLangArgs() {
       case GpuCodegenType::GPU_CG_AMD_HIP:
         args += "hip";
         break;
+      case GpuCodegenType::GPU_CG_CPU:
```
That was my intention first. But I couldn't find a flag that does that in `clang`. The typical way is to use `clang++` instead. I couldn't be certain if that is equivalent to `-x cuda`/`-x hip`, or whether it has other implications. So, I shied away from it.

stonea replied:
Looks like it's valid to do `-x c++` but "hello world" fails unless I also explicitly pass in `-lstdc++`.

If it's easy to do I think it's worth trying to see if we can get this to work with `-x c++` and `-lstdc++` but if there's any hiccups then move on by filing a public GH issue, documenting this caveat in the technote, and considering it a low-priority feature request to get this mode to do c+++ compilation waiting to see if we get any user feedback or discover that it would be useful for our purposes.

e-kayrakli commented on compiler/main/driver.cpp:
```diff
@@ -1555,8 +1553,11 @@ static void setChapelEnvs() {
         gGpuSdkPath = envMap["CHPL_ROCM_PATH"];
         break;
+      case GpuCodegenType::GPU_CG_CPU:
+        gGpuSdkPath = "";
```
Yeah, arguments you have there are the reason why I wanted to stick with it. But I did not draw parallels between this environment and `CHPL_GPU_ARCH`. That's a good point.

Strangely, though, thinking about `CHPL_GPU_ARCH` makes me want to coalesce the path variables instead of creating two different GPU arch variables. I guess I am happy with separate path variables themselves but scared of the slippery slope more.

I will create an issue for this.

e-kayrakli replied:
https://github.com/chapel-lang/chapel/issues/22267

e-kayrakli commented on compiler/main/driver.cpp:
```diff
@@ -1670,14 +1671,16 @@ static void setGPUFlags() {
-      fNoStackChecks  = true;
-      fNoCastChecks = true;
-      fNoDivZeroChecks = true;
+       // if (doGpuCodegen()) {  we still can't do this.
```
> I'm not sure I understand why you added this line and commented it out. Is it something you forgot to remove?

It was so that we don't forget discussing about it :)

Halfway through the implementation, I was somewhat hopeful that we can actually put the checks in with this mode so that you can have an easier debugging experience. But then it didn't work. I guess you can interpret this as something good -- that this mode doesn't do anything too differently compared to "real" gpu modes. By your comment, I would assume that you're in that camp. i.e., You wouldn't like this mode to do bounds checking even if we were able to?

stonea replied:
Yeah my inclination is for this mode to stay as close in behavior to what we have with other GPU modes as possible but your "these would be useful for an easier debugging" experience point seems compelling as well.

But since it didn't work anyway I guess we can avoid trying to think about what the behavior should be for now :-). Of course my hope would be that we could just make these checks part of "normal" GPU support soon.

e-kayrakli commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -169,6 +171,8 @@ class GpuizableLoop {
   std::vector<Symbol*> lowerBounds_;
   bool shouldErrorIfNotGpuizable_;
 
+  std::vector<CallExpr*> gpuAssertions_; // should this be a single CallExpr* ?
```
Good point. To be clear, in that code, you'd want to get a warning from both of these assertions, right? Do you also want that code to be valid if the `proc` wasn't `inline`?

I am not entirely sure if there's an easy way to emit that warning only once per loop, especially for the case where the called function is not inline. Assertions in the loop body are somewhat easier as you can replace them with:


```chapel
var asserted = false;
if !asserted {
  chpl_gpu_assert_on_gpu();
  asserted = true;
}
```

That's not so easy to do for a function that's not inlined, though.

With your general comment about relocating assertions, would you think asserting per iteration but using a `config const` to suppress that warning a better approach?

stonea replied:
> Good point. To be clear, in that code, you'd want to get a warning from both of these assertions, right? Do you also want that code to be valid if the proc wasn't inline?

I'm not sure, I'm starting to have second thoughts if its even worth having it warn it all. It seems like it would be a real nuisance to get the warning multiple times (especially if it becomes once per thread). So I think if we do want to keep `assertOnGpu` and we want it to warn when "passing" with `CHPL_GPU=cpu` then I'd yeah I'd lean towards having a highly advertised `config const` to suppress it.

I still think we might want to consider something like this though:

- Having an `@assertGpuEligible` or if we want to be bold `@assertGpuizable` annotation if the user cares about having a compile-time check.
- Just telling the user to do `assert(here.isGpu())` if they want a runtime check. This would require us to get halting to work for GPUs (or maybe we could just pattern match and remove instances of it). Note that this would just silently pass if `CHPL_GPU=cpu` and the user is on one of the "simulated" gpu sublocales.

e-kayrakli replied:
> So I think if we do want to keep assertOnGpu and we want it to warn when "passing" with CHPL_GPU=cpu then I'd yeah I'd lean towards having a highly advertised config const to suppress it.

I think this is the safest incremental step for this PR.

> I still think we might want to consider something like this though:

I can get behind that proposal. Arguably, this is one facet of replacing the `GPU` modules features with existing Chapel features as much as possible.

e-kayrakli commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -424,6 +432,28 @@ bool GpuizableLoop::extractUpperBound() {
   return true;
 }
 
+void GpuizableLoop::addGpuAssertion(CallExpr* call) {
```
Makes sense. But I'll wait until we settle on the desired behavior for `assertOnGpu` in this mode.

e-kayrakli commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -445,7 +475,7 @@ void GpuizableLoop::reportNotGpuizable(const BaseAST* ast, const char *msg) {
 //      parameters to this new function.
 class GpuKernel {
-  const GpuizableLoop &gpuLoop;
+  GpuizableLoop &gpuLoop;
```
If we do what I am suggesting in https://github.com/chapel-lang/chapel/pull/22172#discussion_r1187952096, this will be moot as this change will be reverted.

e-kayrakli commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -645,6 +675,9 @@ void GpuKernel::populateBody(CForLoop *loop, FnSymbol *outlinedFunction) {
       copyNode = false;
+
+      gpuLoop.addGpuAssertion(call);
+      loop->insertBefore(node->remove());
```
Yeah, it was forgotten in. In any case, the desired behavior is what we're mainly discussing as part of this review. So, I'll wait to make any change here.

e-kayrakli commented on doc/rst/technotes/gpu.rst:
```diff
@@ -140,17 +140,16 @@ before building Chapel.
-``CHPL_GPU_CODEGEN=cuda`` and ``gfx906`` for ``CHPL_GPU_CODEGEN=rocm``. You may
-also use the ``--gpu-arch`` compiler flag to set GPU architecture. For a list
-of possible values please refer to `CUDA Programming Guide
+``CHPL_GPU=nvidia`` and ``gfx906`` for ``CHPL_GPU=amd``. You may also use the
```
I did some more grepping and fixed some stuff.

Note that this PR kinda asks for yet another rename in our nightly tests. But my stance is that we don't touch them just yet to avoid churn. We can revisit when:

- either we make `array_on_device` the default, which will require some shuffling, anyways, or
- when we add intel support.

stonea replied:
Having gone through it once, it's not that difficult to rename the jobs (just somewhat nervewracking). But yeah I agree I'm inclined to let the names sit untouched unless we need to rename them anyway or someone speaks up that they're sufficiently irked by the inconsistencies.

e-kayrakli commented on doc/rst/technotes/gpu.rst:
```diff
@@ -266,6 +265,37 @@ can reduce execution performance significantly, making profiling less valuable.
+  * e.g, ``assertOnGpu`` will fail at compile time normally. This can allow
+    testing if a loop is GPU-eligible.
+
+  * but it will generate only a warning per-loop at execution time.
```
I find both of those points interesting. I'll spawn off an issue to discuss the future of `assertOnGpu`.

e-kayrakli commented on runtime/src/chpl-gpu.c:
```diff
@@ -34,9 +35,41 @@ bool chpl_gpu_debug = false;
+  int32_t num = -1;
+  if ((env = chpl_env_rt_get("NUM_GPUS_PER_LOCALE", NULL)) != NULL) {
+    if (sscanf(env, "%" SCNi32, &num) != 1) {
+      chpl_error("Cannot parse CHPL_RT_NUM_GPUS_PER_LOCALE environment "
```
> I think it's great that your validating this, I don't know if this is the right place to do it or not.

This is heavily inspired by https://github.com/chapel-lang/chapel/blob/main/runtime/src/chpl-tasks.c#L56-L89.

> Is there some common location we use for validating environment variables used by the runtime?

By looking at the code I linked above, my guess is "no".

> Should it be done by one of the chplenv scripts?

I am not sure if that's possible. You can set this env while running your application. I don't think application startup uses those scripts in any meaningful way today.

If anything, this could be something we can ask the `ChplConfig` module to do. I am not sure what other environment variables are there to justify that, though.

e-kayrakli commented on runtime/src/chpl-gpu.c:
```diff
@@ -34,9 +35,41 @@ bool chpl_gpu_debug = false;
+    if (chpl_gpu_num_devices > 0 && num > chpl_gpu_num_devices) {
+      char msg[200];
+      snprintf(msg, sizeof(msg),
+          "CHPL_RT_NUM_GPUS_PER_LOCALE = %" PRIi32 " is too large; "
```
> I could imagine some frankenstein system where some nodes have more GPUs than others 

I *really* don't want to think about such systems. We don't support such systems in general, anyways. e.g., the `CHPL_RT_NUM_THREADS_PER_LOCALE` doesn't check for that case, either.

I don't know how well-documented that assumption is, though. If we were to document it, it should be somewhere more general than the GPU technote, though we may link to that place in the technote. I don't know where that should appear in the docs.

stonea replied:
> I really don't want to think about such systems. We don't support such systems in general, anyways. e.g., the CHPL_RT_NUM_THREADS_PER_LOCALE doesn't check for that case, either.

I'm not sure `CHPL_RT_NUM_THREADS_PER_LOCALE` is really analogous in this case. If `CHPL_RT_NUM_THREADS_PER_LOCALE` exceeds the number of logical or physical cores per nodes that seems fine but if `CHPL_RT_NUM_GPUS_PER_LOCALE` exceeds the number of physical devices (and `CHPL_GPU!=cpu`) then that just seems totally confused.

Would it be that hard to do a quick check across all locales? I think effectively that's what's going to happen anyway once I have this function called by all the nodes in the multilocale PR (https://github.com/chapel-lang/chapel/pull/22258).

e-kayrakli replied:
Actually, the way this is implemented, it is per-locale already. `chpl_gpu_init` must be, and it is, getting called for each locale. If you run this with `-nl 2`, you'll get two warnings if your environment's value exceeds what's available. I added node ID in the message on my local branch:

```
warning: CHPL_RT_NUM_GPUS_PER_LOCALE = 2 is too large; it must be less than or equal to the number of GPUs per node. Detected 1 GPUs in node 0. Ignoring this environment variable.
warning: CHPL_RT_NUM_GPUS_PER_LOCALE = 2 is too large; it must be less than or equal to the number of GPUs per node. Detected 1 GPUs in node 1. Ignoring this environment variable.
on GPU:
1.0 1.20906 2.19525 2.9034 3.5552 4.01125 4.21599 4.10448 3.62297 2.85669 1.60226 1.0
on CPU:
1.0 1.20906 2.19525 2.9034 3.5552 4.01125 4.21599 4.10448 3.62297 2.85669 1.60226 1.0
```

e-kayrakli commented on runtime/include/chpl-mem-array.h:
```diff
@@ -57,7 +57,7 @@ void* chpl_mem_array_alloc(size_t nmemb, size_t eltSize,
   void* p = NULL;
   const size_t size = nmemb * eltSize;
-#ifdef HAS_GPU_LOCALE
+#if defined(HAS_GPU_LOCALE) && !defined(GPU_RUNTIME_CPU)
```
> Though that would add the risk of a super confusing situation where it might not be defined where you need it to be if that common header. I'm not going to push on it, but something to consider.

This scared me. I'll leave it as-is.

e-kayrakli commented:
@stonea -- I believe I have addressed all your feedback. This is ready for another review. We discussed offline but the current behavior for `assertOnGpu` in this mode is that it remains within the loop and triggers a warning per-iteration. You can quiet that warning with `--gpuNoCpuModeWarning` or `CHPL_GPU_NO_CPU_MODE_WARNING`.

stonea commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -169,6 +171,8 @@ class GpuizableLoop {
   std::vector<Symbol*> lowerBounds_;
   bool shouldErrorIfNotGpuizable_;
 
+  std::vector<CallExpr*> gpuAssertions_; // should this be a single CallExpr* ?
```
You can remove this declaration of `gpuAssertions_` now, right? I don't see it used anywhere if I ctrl+F in the diff.

stonea commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -456,7 +461,7 @@ class GpuKernel {
   SymExpr* blockSize_;
 
-  GpuKernel(const GpuizableLoop &gpuLoop, DefExpr* insertionPoint);
+  GpuKernel(GpuizableLoop &gpuLoop, DefExpr* insertionPoint);
```
And we can rake gpuLoop `const` again here, right?

stonea commented on modules/internal/ChapelGpuSupport.chpl:
```diff
@@ -23,16 +23,19 @@ module ChapelGpuSupport {
   config const debugGpu = false;
 
+  extern var chpl_gpu_no_cpu_mode_warning: bool;
+  config const gpuNoCpuModeWarning = isEnvSet("CHPL_GPU_NO_CPU_MODE_WARNING");
```
There's something about `CHPL_GPU_NO_CPU_MODE_WARNING` that makes it seem different from other `CHPL_` variables:

* It doesn't get printed out by `printchplenv --all`.
* I don't think it would get addressed if put in a Chapel configuration file (https://chapel-lang.org/docs/usingchapel/chplenv.html#chapel-configuration-files).
* It's strictly a runtime configuration thing that doesn't impact the build of either the compiler or the runtime library.

I suppose `CHPL_DEBUG` doesn't fit these first two bullets. We also have `CHPL_RT_` variables, is this more like one of those?

Do we have a precedent for other `CHPL_` variables that impact runtime behavior / are used to assign values to config consts like this?

I'm not opposed to having this as a `CHPL_` variable but this seems different enough from what we do with other `CHPL_` variables I'm familiar with that I question if this is somehow falling out of our overall design philosophy for these. If we have a precedent of doing this ignore my comment but otherwise you might want to get a second opinion.

e-kayrakli replied:
> - It doesn't get printed out by printchplenv --all
> - I don't think it would get addressed if put in a Chapel configuration file (https://chapel-lang.org/docs/usingchapel/chplenv.html#chapel-configuration-files).

Not all environment variables (either `CHPL_` or `CHPL_RT_`) we care about do. Note that for most (all?) compiler flags, there's a corresponding environment variable that we take a look, as well. So, we don't try to be exhaustive with respect to those bullets.

> Do we have a precedent for other CHPL_ variables that impact runtime behavior / are used to assign values to config consts like this?

Maybe not `config const`, but here's an example that I am aware and based this off of:
https://github.com/chapel-lang/chapel/blob/main/modules/packages/CopyAggregation.chpl#L95-L97

> I'm not opposed to having this as a CHPL_ variable but this seems different enough from what we do with other CHPL_ variables I'm familiar with that I question if this is somehow falling out of our overall design philosophy for these. If we have a precedent of doing this ignore my comment but otherwise you might want to get a second opinion.

That's a valid question, but based on my responses above; I am not bothered.

---

OTOH,

> We also have CHPL_RT_ variables, is this more like one of those?

This is true. Maybe `CHPL_RT_` is a more consistent prefix for this. If you prefer that, I am happy to make that change.

stonea replied:
> Not all environment variables (either CHPL_ or CHPL_RT_) we care about do. Note that for most (all?) compiler flags, there's a corresponding environment variable that we take a look, as well. So, we don't try to be exhaustive with respect to those bullets.

Ok cool. I just wanted to check that we weren't going "out of model" with this. I'd be fine having with either as a `CHPL_` or `CHPL_RT_` variable then.

e-kayrakli replied:
Leaving as-is.

stonea commented on modules/internal/ChapelGpuSupport.chpl:
```diff
@@ -42,4 +45,10 @@ module ChapelGpuSupport {
             if canAccessPeer(gpu1,gpu2) then setPeerAccess(gpu1,gpu2,true);
     }
+
+  private proc isEnvSet(name: string): bool {
```
This seems sufficiently general that I'd be surprised if we didn't already have something in our standard modules that did it.

It looks like you could do this by looking at the return value of: `Sys.sys_getenv`

https://chapel-lang.org/docs/modules/standard/Sys.html?highlight=getenv#Sys.sys_getenv

Although the comment there tells you that, that's deprecated and to use: `OS.sys_getenv`.

Funny enough if I look at the doc for the `OS` module (https://chapel-lang.org/docs/modules/standard/OS.html) I don't see `sys_getenv` documented there. So maybe its no doc'd?

Anyway, you might want to see if you could use `OS.sys_getenv` instead and could you poke whoever is responsible for stabilizing the sys and/or OS module? Seems like we have an issue in our doc about it if nothing else.

e-kayrakli replied:
Good catch! For posterity: I brought this to the attention of the 2.0 subteam. I'll click resolve on this convo as there's not much I can do at the moment. But I am hoping that this will be fixed once we have a standard way of reading an environment variable.

stonea commented on modules/standard/GPU.chpl:
```diff
@@ -189,8 +190,15 @@ module GPU
+      const voidPtr = __primitive("gpu allocShared", numBytes(eltType)*size);
+      return voidPtr : c_ptr(eltType);
+    }
+    else {
```
Any reason not to do this in the runtime instead? `__primitive("gpu allocShared")` just calls out to a runtime function that we have different `impl` functions for. Couldn't you just have that call `chpl_malloc` or whatever?

I don't know if it's always possible, but I think it's preferable if we can get as much of our vendor-specific code into the vendor specific runtime libs. It also has the added benefit that things would still work if the primitive (for whatever reason) was called directly (either explicitly in Chapel code or if we code gen'd it somewhere else for whatever reason).

e-kayrakli replied:
> `__primitive("gpu allocShared")` just calls out to a runtime function that we have different impl functions for.

Does it? It doesn't look like it https://github.com/chapel-lang/chapel/blob/main/compiler/codegen/cg-expr.cpp#L5268

I think that should be doable. If that was the case, I agree with your sentiment here. But I am not willing to add that to this PR.

stonea replied:
Ah you're right, I must have been mixing this up with what we do for sync threads. Ok, I'm fine with how it is in your PR.

stonea commented on runtime/include/chpl-mem.h:
```diff
@@ -167,7 +167,8 @@ void* chpl_memcpy(void* dest, const void* src, size_t num)
 void* chpl_memmove(void* dest, const void* src, size_t num)
 {
-#ifdef HAS_GPU_LOCALE
+  // check for GPU_LOCALE_NONE is needed to break recursion
```
`GPU_LOCALE_NONE` should read `GPU_RUNTIME_CPU` here, right?

stonea commented on doc/rst/technotes/gpu.rst:
```diff
@@ -139,22 +139,27 @@ To enable GPU support set the environment variable ``CHPL_LOCALE_MODEL=gpu``
-different installation) you may set ``CHPL_CUDA_PATH`` and/or
-``CHPL_ROCM_PATH``.
+have and where your installation of relevant runtime (e.g. CUDA or ROCm) are.
+If the type of GPU is not detected you may set ``CHPL_GPU`` manually to either
```
If the type of GPU is not detected you may set ``CHPL_GPU`` manually to either ``nvidia`` or ``amd``

-->  

If the type of GPU is not detected you may set ``CHPL_GPU`` manually to either ``nvidia``, ``amd``, or ``cpu`` (see here [link to cpu section]).

stonea commented on doc/rst/technotes/gpu.rst:
```diff
@@ -139,22 +139,27 @@ To enable GPU support set the environment variable ``CHPL_LOCALE_MODEL=gpu``
+sublocales created. For example, in a system where 8 physical GPUs per node,
+setting this environment variable to 4 will make Chapel use only the first 4
+GPUs. Setting this number to a value higher than the number of GPUs per node
+results in an error.
```
Add: When `CHPL_GPU=cpu`, you may set `CHPL_RT_NUM_GPUS_PER_LOCALE` to any non-negative number.

stonea reviewed (approved):
Ok, I've taken my second pass and added a handful of additional comments. Feel free to address what you wish.

stonea commented on compiler/main/driver.cpp:
```diff
@@ -1555,8 +1553,11 @@ static void setChapelEnvs() {
+      case GpuCodegenType::GPU_CG_CPU:
+        gGpuSdkPath = "";
+        break;
       default:
```
Right, I guess I'm pointing things out that are adjacent to but not strictly part of your PR.

My understanding is if you don't have the `default` case and we added a new `GpuCodegenType` (say `GPU_CG_INTEL`) then when you try and compile Chapel you'll get an error or warning about that missing case and it'd be better to discover this as a compiler error than a warning.

That said, I feel like I've seen this pattern of using `INT_FATAL` as the default case in other enums that we intend to exhaustively cover so I'm paranoid that I'm telling you something wrong.

e-kayrakli replied:
I am with you. I don't know if there's a deep reason why we fall into `INT_FATAL` rather then relying on a build time error for a missing case. Maybe the behavior for a missing case is undefined in C++ and we don't want to rely on compilers' behavior?? I'll look around a bit.

e-kayrakli replied:
Looks like we do exactly what you suggest in 2 other similar locations. I dropped the default case.

bradcray commented:
Nice!
